Домашнее задание №5 «Подсчет площади геометрических фигур»

chepsel opened the pull request:
- [x] Создайте интерфейс Shape, который будет иметь метод для вычисления площади.
- [x] Реализуйте структуры для геометрических фигур: круг, прямоугольник, треугольник. Каждая структура должна реализовывать интерфейс Shape.
- [x] Создайте функцию calculateArea(s any), которая ожидает на входе объект типа Shape и возвращает его площадь. Если переданный объект не реализует интерфейс Shape, функция должна вернуть ошибку.
- [x] В функции main создайте несколько объектов разных типов (круг, прямоугольник, треугольник) и вызвать функцию calculateArea для каждого из них. Обработайте возможные ошибки и выведите результаты подсчета площадей на экран.

chepsel committed:
5й дз - попытко один

chepsel committed:
Merge branch 'master' into hw05_shapes

hungry-honeybadger commented on hw05_shapes/main.go:
```diff
@@ -1,5 +1,96 @@
+}
+
+type Triangle struct {
+	base   uint16
```
Экспортируемая структура с неэкспортируемыми полями. Данная структура не имеет смысла . Так как в другом пакете мы не будем иметь возможности установить значение полей base и height. Для решения этой проблемы нужно либо добавить сетеры, либо конструктор. Я бы рекомендовал все таки добавить конструктор, так как высота и основание у треугольника это неизменяемые значения( если их изменить это будет уже другой треугольник)

chepsel replied:
Не подумал, спасибо!) 
Поправил

hungry-honeybadger reviewed (changes requested):
Хорошее начало, но код имеет явную рассогласованность экспоритруемости. Для того, что бы понять почему это так рекомендую вынести круг, прямоугольник и квадрат в отдельный пакет. При этом оставив интерфейс в пакете main( так как в go принято хранить интерфейсы так, где они используются) и попробовать запустить этот код

chepsel committed:
5й дз - попытко два

chepsel committed:
test
Merge branch 'hw05_shapes' of https://github.com/chepsel/home_work_basic into hw05_shapes

hungry-honeybadger commented on hw05_shapes/figure/pkg.go:
```diff
@@ -0,0 +1,70 @@
+	return 0, fmt.Errorf("circle input parameter is 0, radius: %v", c.radius)
+}
+
+func NewTriangle(b uint16, h uint16) *Triangle {
```
Аргументам нужно давать нормальные, понятные имена.

hungry-honeybadger commented on hw05_shapes/figure/pkg.go:
```diff
@@ -0,0 +1,70 @@
+	}
+}
+
+func NewRectangle(w uint16, h uint16) *Rectangle {
```
Аргументам нужно давать нормальные, понятные имена.

hungry-honeybadger commented on hw05_shapes/figure/pkg.go:
```diff
@@ -0,0 +1,70 @@
+	}
+}
+
+func NewCircle(r float64) *Circle {
```
Аргументам нужно давать нормальные, понятные имена.

hungry-honeybadger commented on hw05_shapes/figure/pkg.go:
```diff
@@ -0,0 +1,70 @@
+	return 0, fmt.Errorf("one of rectangle input parameters is 0, width: %v height %v", r.width, r.height)
+}
+
+func (c *Circle) GetArea() (any, error) {
```
Функция должна возвращать float, а не any.

hungry-honeybadger commented on hw05_shapes/figure/pkg.go:
```diff
@@ -0,0 +1,70 @@
+	return 0, fmt.Errorf("one of triangle input parameters is 0, base: %v height %v", t.base, t.height)
+}
+
+func (r *Rectangle) GetArea() (any, error) {
```
Функция должна возвращать float, а не any.

hungry-honeybadger commented on hw05_shapes/figure/pkg.go:
```diff
@@ -0,0 +1,70 @@
+	return 1
+}
+
+func (t *Triangle) GetArea() (any, error) {
```
Функция должна возвращать float, а не any.

hungry-honeybadger commented on hw05_shapes/main.go:
```diff
@@ -1,5 +1,69 @@
+)
+
+type Shape interface {
+	GetArea() (any, error)
```
Функция должна возвращать float, а не any.

hungry-honeybadger commented on hw05_shapes/main.go:
```diff
@@ -1,5 +1,69 @@
+	GetBeer() any
+}
+
+func calculateArea(s any) (any, error) {
```
Функция должна возвращать float, а не any.

hungry-honeybadger commented on hw05_shapes/main.go:
```diff
@@ -1,5 +1,69 @@
+func calculateArea(s any) (any, error) {
+	switch t := s.(type) {
+	case Shape:
+		return s.(Shape).GetArea()
```
Тут лучше обратится сразу к t.

hungry-honeybadger reviewed (changes requested):
Работа стала значительно лучше. Но площадь это число, поэтому функция вычисляющая ее должна возвращать число

chepsel committed:
5й дз - попытко три

chepsel committed:
5й дз - попытко 4(обработал пропущенный коммент)
